addons/mail/models/mail_activity.py

```diff
@@ -571,53 +584,160 @@ def action_open_document(self):
+        self.attachment_ids.fetch(['name'])
         for record, activity in zip(self, activities):
             activity['mail_template_ids'] = [
                 {'id': mail_template.id, 'name': mail_template.name}
                 for mail_template in record.mail_template_ids
             ]
+            activity['attachment_ids'] = [
+                {'id': attachment.id, 'name': attachment.name}
+                for attachment in record.attachment_ids
+            ]
         return activities
 
     @api.model
-    def get_activity_data(self, res_model, domain, limit=None, offset=0):
+    def get_activity_data(self, res_model, domain, limit=None, offset=0, fetch_done=False):
+        """ Get aggregate data about records and their activities.
+
+        The goal is to display aggregated data about records and their activities for
+        example in activity views and the chatter.
+
+        The aggregated data returned are based on the records along with their related
+        activities that are fetched according to the search criterion: res_model,
+        domain, limit, offset and fetch_done. The fetch_done parameter determines if
+        "done" activity are integrated in the aggregated data or not.
+
+        :param str res_model: model of the records to fetch
+        :param list domain: record search domain
+        :param int limit: maximum number of records to fetch
+        :param int offset: offset of the first record to fetch
+        :return dict: {'activity_types': list of activity type info as list
+                            (id, name, mail_template, keep_done),
+                       'activity_res_ids': list<int> of record id ordered by closest date
+                            (deadline for ongoing activities, and done date for done activities),
+                       'grouped_activities': dict<dict<dict>>
+                            res_id -> activity_type_id -> dict of aggregated info
+                            where the aggregate information structured as:
+                                count_by_state: dict mapping state to count (ex.: 'planned': 2)
+                                ids: list of activity ids for the res_id and activity_type_id
+                                o_closest_date: closest date (deadline if ongoing, done date otherwise)
+                                state: aggregated state of the activities
+                                user_ids_ordered_by_deadline: activity responsible ordered by closest
+                                    deadline of the related activity
+                       }
+        """
+        user_tz = self.user_id.sudo().tz
+        Model = self.env[res_model]
+
+        # 1. Retrieve all ongoing and completed activities according to the parameters
+        activity_types = self.env['mail.activity.type'].search([('res_model', 'in', (res_model, False))],
+                                                               order='sequence')
```

Comment:
just to have something determinist if some have the same sequence

```suggestion
                                                               order='sequence, id')

Reply:
Yes, good idea